From afb955026c84d498d9330abf28cd6180a5406b11 Mon Sep 17 00:00:00 2001
From: Xu Liang <liang.xu@intel.com>
Date: Sat, 19 Sep 2020 00:03:15 +0800
Subject: [PATCH] DRVLIB_SW-3883: GSW API - PAE early detection of routing
 session table full based on number of used (valid) entries

---
 drivers/net/ethernet/lantiq/switch-api/gsw_pae.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/drivers/net/ethernet/lantiq/switch-api/gsw_pae.c b/drivers/net/ethernet/lantiq/switch-api/gsw_pae.c
index 4697b8ed982f..cc6e0717cc36 100644
--- a/drivers/net/ethernet/lantiq/switch-api/gsw_pae.c
+++ b/drivers/net/ethernet/lantiq/switch-api/gsw_pae.c
@@ -708,6 +708,11 @@ int GSW_ROUTE_SessionEntryAdd(void *cdev, GSW_ROUTE_Entry_t *rpar)
 
 
 	if ((ethdev->gipver == LTQ_GSWIP_3_0)	&& ((ethdev->gsw_dev == LTQ_FLOW_DEV_INT_R))) {
+		if (rthandler->rstbl.nuentries >= RT_SESSION_TABLE_SIZE) {
+			retval = GSW_ROUTE_ERROR_RT_SESS_FULL;
+			goto UNLOCK_AND_RETURN;
+		}
+
 		if (rpar->nHashVal < 0) {
 			/* calculate hash index */
 			hindex = pae_hash_index(rpar);
